File: TwoPointers/MoveZeroes.py
class Solution(object):
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """

        # 'left' points to the next position where a non-zero element
        # should be placed, while 'right' scans the array.
        left = 0
        right = 0
        end = len(nums) - 1

        # Move all non-zero elements to the front of the array.
        while right <= end:

            # If the current element is non-zero,
            # place it at the 'left' pointer.
            if nums[right] != 0:
                nums[left] = nums[right]
                right += 1
                left += 1

            # Otherwise, simply continue scanning.
            else:
                right += 1

        # Fill the remaining positions with zeros.
        while left <= end:
            nums[left] = 0
            left += 1

        return nums

File: app/content/solutions/move-zeroes.c
/* Translated from dsa-notes/TwoPointers/MoveZeroes.py — same two-pointer
   algorithm, written the way the exam expects. */
#include <stdio.h>

int main(void) {
    int n;
    if (scanf("%d", &n) != 1) { return 0; }

    int nums[1000];
    for (int i = 0; i < n; i++) {
        scanf("%d", &nums[i]);
    }

    /* 'left' is where the next non-zero belongs; 'right' scans. */
    int left = 0;
    for (int right = 0; right < n; right++) {
        if (nums[right] != 0) {
            nums[left] = nums[right];
            left++;
        }
    }

    /* Fill the tail with zeroes. */
    while (left < n) {
        nums[left] = 0;
        left++;
    }

    for (int i = 0; i < n; i++) {
        printf("%d", nums[i]);
        if (i < n - 1) printf(" ");
    }
    printf("\n");
    return 0;
}
